data/basics/overflow.go

```diff
@@ -75,94 +55,60 @@ func OMul(a uint64, b uint64) (res uint64, overflowed bool) {
 }
 
 // MulSaturate multiplies 2 values with saturation on overflow
-func MulSaturate(a uint64, b uint64) uint64 {
+func MulSaturate[T constraints.Unsigned](a, b T) T {
 	res, overflowed := OMul(a, b)
 	if overflowed {
-		return math.MaxUint64
+		var defaultT T
+		return ^defaultT
 	}
 	return res
 }
 
 // AddSaturate adds 2 values with saturation on overflow
-func AddSaturate(a uint64, b uint64) uint64 {
+func AddSaturate[T constraints.Unsigned](a, b T) T {
 	res, overflowed := OAdd(a, b)
 	if overflowed {
-		return math.MaxUint64
-	}
-	return res
-}
-
-// AddSaturate32 adds 2 uint32 values with saturation on overflow
-func AddSaturate32(a uint32, b uint32) uint32 {
-	res, overflowed := OAdd32(a, b)
-	if overflowed {
-		return math.MaxUint32
+		var defaultT T
+		return ^defaultT
 	}
 	return res
 }
 
 // SubSaturate subtracts 2 values with saturation on underflow
-func SubSaturate(a uint64, b uint64) uint64 {
+func SubSaturate[T constraints.Unsigned](a, b T) T {
 	res, overflowed := OSub(a, b)
 	if overflowed {
 		return 0
 	}
 	return res
 }
 
-// SubSaturate32 subtracts 2 uint32 values with saturation on underflow
-func SubSaturate32(a uint32, b uint32) uint32 {
-	res, overflowed := OSub32(a, b)
-	if overflowed {
-		return 0
-	}
-	return res
-}
-
-// Add16 adds 2 uint16 values with overflow detection
-func (t *OverflowTracker) Add16(a uint16, b uint16) uint16 {
-	res, overflowed := OAdd16(a, b)
-	if overflowed {
-		t.Overflowed = true
-	}
-	return res
-}
-
 // Add adds 2 values with overflow detection
-func (t *OverflowTracker) Add(a uint64, b uint64) uint64 {
-	res, overflowed := OAdd(a, b)
-	if overflowed {
-		t.Overflowed = true
-	}
-	return res
+func (t *OverflowTracker) Add(a, b uint64) (res uint64) {
+	res, t.Overflowed = OAdd(a, b)
```

Comment:
I think it's important to only assign true values to `t.Overflowed`, since `OverflowTracker` (I believe) is meant to be used with multiple operations.

With this change, as long as later operations don't overflow, you will forget if earlier ones do.

Reply:
Excellent catch. Let's make sure we have a unit test to catch this too.

Reply:
see [revert the stupid compression commit](https://github.com/algorand/go-algorand/pull/5501/commits/b3c093f441f73837f9f33f2e363f10fc96062f1b)

Reply:
Right, looks like this is tested in `TestNextRewardsRateWithFixRewardsWithResidueOverflow`, so I'll resolve